### DISCUSSION TITLE
setTX - setRX causes board COM to disappear

### DISCUSSION THREAD

**Discussion by dBarmpxkos** [earlephilhower/arduino-pico#2682] [General]
I am probing rp2040 GP4 for output using the sketch below and I am running in an issue.

```
#include <MIDI.h>

MIDI_CREATE_INSTANCE(HardwareSerial, Serial2, TRS_MIDI);

void setup() {
  Serial2.setTX(4);
  Serial2.setRX(5);
  TRS_MIDI.begin(MIDI_CHANNEL_OMNI);
  Serial2.begin(31250);
}

void loop() {
  TRS_MIDI.sendNoteOn(40, 128, 0);
  delay(500);
}
```

After I flash the device with this code, the COM port stops appearing and the only way to reprogram it is by putting in into bootloader mode so it can appear as UF2 board again! The sketch works as intended, I can validate it with the oscilloscope.

**Comment by elral** [comment]
I also run into this issue today. I wanted to remap the pins for Serial1 and Serial 2, but after flashing the Serial COM port does not appear again.
Is there a reason why it was converted into a discussion?

**Comment by elral** [comment]
I had a short look into this. I wasn't aware that only defined pins can be used. But even with this pins the COM port doesn't appear.
It looks like changing the pin defininition in pins_arduino.h does work, but I have to double check this.

**Reply by dBarmpxkos** [reply]
I have tested a bare minimum example, just changing serial pins to ones that are legally allowed.  After uploading the sketch, COM port disappears from my PC.

**Reply by earlephilhower** [reply]
This could be related to your board definition.  Some boards "helpfully" redefine `Serial1` and `Serial2`.  If you build and run on the standard Pico then the minimum script above runs just fine and dumps X to USB:

````
void setup() {
  Serial2.setTX(4);
  Serial2.setRX(5);
  Serial2.begin(31250);
}

void loop() {
  delay(500);
  Serial.printf("x\n");
}
````

If the board manufacturer renamed the ports, then just try `Serial1` instead of `Serial2`.  You can also hook up SWD and see what's going on.

**Comment by elral** [comment]
It was my own stupid failure
```
Serial1.setTX(SERIAL1_TX_PIN);
Serial1.setRX(SERIAL1_RX_PIN);
Serial1.setTX(SERIAL2_TX_PIN);
Serial1.setRX(SERIAL2_RX_PIN);
```
Last two commands should be for Serial**2**
I should be more carefully with copy and paste :(
But interesting that using a wrong pin causes not appearing the serial connection via USB.
Sorry for that.

**Comment by Phozer** [comment]
I am having the same Issue. I want to use USB Serial for debugging. Serial1 as UART0 GPIO 0&1 and Serial 2 as UART1 GPIO 6&7. As soon i enabled Serial2 the COM Port disapears and i need to unplug the pico, press bootsel to upload a new firmware manually via draging the uf2 file again. What is causing this Issue? Help is appreciated. 
```c++
// Raspberry Pi Pico
// UART0 Pins
const int TX0_PIN = 0; // GPIO 0 as TX
const int RX0_PIN = 1; // GPIO 1 as RX

// UART1 Pins
const int TX1_PIN = 6; // GPIO 6 as TX
const int RX1_PIN = 7; // GPIO 7 as RX

void sendPacket(SerialUART &uart, uint8_t command, uint8_t *data, uint8_t dataLength);
void processIncomingPacket(SerialUART &uart);
void decodeCommand(uint8_t command, uint8_t *data, uint8_t dataLength);

void setup()
{
  // Configure UART0 pins
  Serial1.setTX(TX0_PIN);
  Serial1.setRX(RX0_PIN);
  // Initialize Serial1 with the required UART settings
  Serial1.begin(19200, SERIAL_8E1);

  // Configure UART1 pins
  Serial2.setTX(TX1_PIN);
  Serial2.setRX(RX1_PIN);
  // Initialize Serial2 with the required UART settings
  Serial2.begin(19200, SERIAL_8E1);

  // Initialize built-in LED
  pinMode(LED_BUILTIN, OUTPUT);
  digitalWrite(LED_BUILTIN, LOW);

  // Initialize USB Serial for debugging
  Serial.begin(115200);
  while (!Serial)
  {
    // Wait for the Serial Monitor
  }
....
```

**Reply by PontusO** [reply]
Pins 6 and 7 are not valid uart tx and rx pins. You probably want pins 8 and 9 instead. The data sheet, section 1.4.3. GPIO Functions, lists what combinations that are valid.

**Reply by Phozer** [reply]
🤦‍♂️😅 Thanks for the lift. I have mistaken the Pin Number with the GPIO-Number. 🙏
